data/tensor.lean

```diff
@@ -0,0 +1,330 @@
+import data.list.basic
+import algebra.module
+import tactic.interactive
+import tactic.tidy
+
+open list
+
+def tensor_index (ds: list ℕ) := { is : list ℕ // forall₂ (<) is ds}
+
+namespace tensor_index
+
+variables {ds₁ : list ℕ} {ds₂ : list ℕ} {ds₃ : list ℕ}
+
+def take : Π {ds₁ : list ℕ}, tensor_index (ds₁ ++ ds₂) → tensor_index ds₁
+| ds is := ⟨ list.take (length ds) is.1, forall₂_take_append is.1 ds ds₂ is.2 ⟩
+
+def drop : Π {ds₁ : list ℕ}, tensor_index (ds₁ ++ ds₂) → tensor_index ds₂
+| ds is := ⟨ list.drop (length ds) is.1, forall₂_drop_append is.1 ds ds₂ is.2 ⟩
+
+lemma cast_type (is : list ℕ) (eq : ds₁ = ds₂) (h: forall₂ (<) is ds₁) :
+  (cast (congr_arg tensor_index eq) ⟨is, h⟩).val = is :=
+by subst eq; refl
+
+def assoc_right :
+  tensor_index (ds₁ ++ ds₂ ++ ds₃) → tensor_index (ds₁ ++ (ds₂ ++ ds₃)) :=
+cast (congr_arg tensor_index (append_assoc ds₁ ds₂ ds₃))
+
+def assoc_left :
+  tensor_index (ds₁ ++ (ds₂ ++ ds₃)) → tensor_index (ds₁ ++ ds₂ ++ ds₃) :=
+cast (congr_arg tensor_index (append_assoc ds₁ ds₂ ds₃).symm)
+
+lemma take_take :
+  ∀ t : tensor_index (ds₁ ++ ds₂ ++ ds₃),
+  t.assoc_right.take = t.take.take
+| ⟨ is , h ⟩ := subtype.eq (by simp [assoc_right,take, cast_type, list.take_take, nat.le_add_right, min_eq_left])
+
+lemma drop_take :
+  ∀ t : tensor_index (ds₁ ++ ds₂ ++ ds₃),
+  t.assoc_right.drop.take = t.take.drop
+| ⟨ is , h ⟩ := subtype.eq (by simp [assoc_right, take, drop, cast_type, list.drop_take])
+
+lemma drop_drop :
+  ∀ t : tensor_index (ds₁ ++ ds₂ ++ ds₃),
+  t.assoc_right.drop.drop = t.drop
+| ⟨ is , h ⟩ := subtype.eq (by simp [assoc_right,drop, cast_type, list.drop_drop])
+
+end tensor_index
+
+def tensor (α : Type) (ds:list ℕ) := tensor_index ds -> α
+
+namespace tensor
+
+variables {α : Type} {d : ℕ} {ds : list ℕ} {ds₁ : list ℕ} {ds₂ : list ℕ} {ds₃ : list ℕ}
+
+instance [inhabited α] : inhabited (tensor α ds) := ⟨λ t, default α⟩
+instance [has_zero α] : has_zero (tensor α ds) := ⟨λ t, 0⟩
+instance [has_add α] : has_add (tensor α ds) := ⟨λ x y t, (x t) + (y t)⟩
+instance [has_neg α] : has_neg (tensor α ds) :=  ⟨λ a t, - a t⟩
+
+instance [add_semigroup α] : add_semigroup (tensor α ds) :=
+  ⟨(+), λ a b c, funext (λt, add_assoc (a t) (b t) (c t))⟩
+
+instance [add_comm_semigroup α] : add_comm_semigroup (tensor α ds) :=
+  ⟨(+), add_assoc, λ a b, funext (λt, add_comm (a t) (b t))⟩
+
+instance [add_monoid α] : add_monoid (tensor α ds) :=
+  ⟨(+), add_assoc, 0, λ a, funext (λt, zero_add (a t)), λ a, funext (λt, add_zero (a t))⟩
+
+instance [add_comm_monoid α] : add_comm_monoid (tensor α ds) :=
+  ⟨(+), add_assoc, 0, zero_add, add_zero, add_comm ⟩
+
+instance [add_group α] : add_group (tensor α ds) :=
+  ⟨(+), add_assoc, 0, zero_add, add_zero, has_neg.neg, λ a, funext (λt, add_left_neg (a t))⟩
+
+instance [add_comm_group α] : add_comm_group (tensor α ds) :=
+  ⟨(+), add_assoc, 0, zero_add, add_zero,  has_neg.neg, add_left_neg, add_comm ⟩
+
```

Comment:
Does pi_instances do the last 20 or so lines in one line, or will it not handle it?

Reply:
Yes, pi_instances can do it. I updated the file.